### PR TITLE
Bulk Load CDK: Exception Handling for TaskRunner failures.

### PR DESCRIPTION
**Pull request by johnny-schmidt**
## What
Adds the basis for recovery from exceptions.

* Exceptions caught in the TaskRunner trigger queue flush + rethrow
* These get caught in the WriteOperation, which restarts the task runner and calls TaskLauncher::handleException. It is assumed the launcher will initiate cleanup tasks
* There is no path to recovery -- it's assumed if the errors were recoverable then the tasks would have recovered.

Next steps: add the actual workflow to the DestinationTaskLauncher. (Keeping the PRs small.)

Note: This was deceptively simple, as there were two big gotchas:

* Because the unit test calls into something that calls runBlocking (WriteOperation::execute), it's possible for the write operation to block the test code; hence the use of Dispatchers.IO. IIRC @stephane-airbyte you said there was an alternative equivalent for use in unit tests? I searched but couldn't find anything.
* `consumeAsFlow()` is not reentrant. It sets up one consumer for life. So attempting to restart the runner would block. `receiveAsFlow()` allows multiple consumers/reattaching, which works for now. Possibly an improvement is to enforce this? There's still only on receiver per channel element, so worst case we have two threads pulling tasks down and executing them.

### CODE REVIEW

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/TaskRunner.kt [hunk 1]
```diff
@@ -31,11 +31,19 @@ class TaskRunner {
...
+        try {
+            queue.receiveAsFlow().collect { task ->
+                launch {
+                    log.info { "Executing task: $task" }
+                    task.execute()
+                }
             }
+        } catch (t: Throwable) {
+            log.error { "Exception in task runner, flushing queue: $t" }
+            queue.receiveAsFlow().collect { task ->
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/TaskRunner.kt:43]: I'm realizing it's not clear to me how this manages to exit without a `close`. I wonder if that means the previous `receiveAsFlow` will stop after emptying the queue?

Comment (stephane-airbyte) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/TaskRunner.kt:43]: receiveAsFlow exits when the underlying channel is closed. I don't think I understand your question

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt [hunk 2]
```diff
@@ -23,18 +23,29 @@ import kotlinx.coroutines.runBlocking
...
     override fun execute() {
         runBlocking {
             launch { inputConsumer.run() }
 
             launch { taskLauncher.start() }
 
-            launch { taskRunner.run() }
+            launch {
+                try {
+                    taskRunner.run()
```
Comment (stephane-airbyte) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt:40]: I do not understand this at all. The double wrapping of `taskRunner.run()` inside 2 different `launch` seems very suspicious. Even the need to call the `run` twice seems odd. Clearly, I don't understand the intent or the logic of the underlying constructs

**Comment by stephane-airbyte** [comment]
> Because the unit test calls into something that calls runBlocking (WriteOperation::execute), it's possible for the write operation to block the test code; hence the use of Dispatchers.IO. IIRC @stephane-airbyte you said there was an alternative equivalent for use in unit tests? I searched but couldn't find anything.

I don't remember any conversation around testing of suspend functions. It's not something I'm familiar with so I can't really provide any insight on howto deal with concurrency issue when testing those constructs

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt [hunk 3]
```diff
@@ -23,18 +23,29 @@ import kotlinx.coroutines.runBlocking
...
+                } catch (t: Throwable) {
+                    launch { taskRunner.run() }
```
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt:42]: @johnny-schmidt is the intent here to continue executing other streams' tasks, even if one stream encounters an error? (I'm guessing wildly, this is basically my first time looking at this region of code)

if yes: the `.run()` inside the catch clause probably also needs to have the equivalent catch+rerun logic?

**Review by edgao — COMMENTED** [review]
I don't think I have the knowledge/context to review this usefully, just trying to get more familiar with these pieces :P so my comments could be completely off-base, feel free to ignore them!

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt (hunk 3 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt:42]: The point is to enable the launcher to dispatch tasks that cleanup after failure, not to keep things running. So a failure at this stage would be a hard failure.

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt (hunk 2 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt:40]:
```
try {
  launch { taskRunner.start() }
catch(t: Throwable) {
  launch { taskRunner.restart() }
  taskLauncher.handleException(t)
}
```
This is probably clearer, but I'm not 100% sure that it works as expected. I'll test it.

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/TaskRunner.kt (hunk 1 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/TaskRunner.kt:43]: I mean that nothing closes the channel during flush, so it should block, but it doesn't. So I'm missing something.

**Comment by johnny-schmidt** [comment]
@edgao @stephane-airbyte 

I resolved the issues flagged above, this is ready for final review.

Exception handling with coroutines is a little tricky, but the following are true

* in normal `coroutineScope`s, any exception cancels the scope and all the child/sibling tasks, so if you want to handle exceptions effectively, you need to wrap all the suspend functions in `coroutineScope` and put your try/catch outside of it. (This is why `receiveAsFlow()` wasn't blocking, its context was canceled by the exception before it could finish flushing the queue.)
* there's a special scope called `supervisorScope` where this is not true. In this scope, any async task can fail independently. use this when you want custom exception handling. (ie, `val job = launch { ... }` then later `if (job.isCancelled()) { ... }` etc.). We're not using it now, but we might in the future. Sources uses it.

Grokking that led to the following changes

* in `WriteOperation`, I put all the launched actions into their own scope, so that they fail each other
* inside `TaskRunner`, I wrap the `receiveAsFlow` call in its own scope, then put the try/catch around that, drain the queue on error and rethrow to the caller.
* (I also added locking semantics to prevent anyone from writing to the queue until the launcher is restarted)

Hopefully the flow should be more clear now.

**Comment by johnny-schmidt** [comment]
> @edgao @stephane-airbyte
> 
> I resolved the issues flagged above, this is ready for final review.
> 
> Exception handling with coroutines is a little tricky, but ...

@rodireich If you have a minute, could you vet the statements in the above comment ^? This PR was tricky.

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt (hunk 2 above)
Comment (stephane-airbyte) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/WriteOperation.kt]: I still don't understand why we need to restart the task

**Review by edgao — COMMENTED** [review]
don't block on me, I don't think I should approve / reject this pr. (that's a thing I want to spend today/monday catching up on)

**Comment by stephane-airbyte** [comment]
@johnny-schmidt As I've communicated, I won't be able to do a proper review of this code, as it depends on a lot of code that I already don't understand. I do really think the total complexity is too much for my understanding. I will mark this as "approved", but it is not a signal that I was able to perform an extensive review of the suggested changes. The approval is more of a sign that I trust you understand the scope of your changes

**Review by stephane-airbyte — APPROVED** [review]
blanket approval based on the trust that authors understand the ins and out of your changes

**Comment by rodireich** [comment]
> in normal coroutineScopes, any exception cancels the scope and all the child/sibling tasks
That's right. the entire operation is cancelled.

> there's a special scope called supervisorScope where this is not true. In this scope, any async task can fail independently. use this when you want custom exception handling.
IIRC in `supervisorScope` one stack of suspend coroutines doesn't affect another stack.
By stack I mean coroutine launching another coroutines in the same scope.
an Exception will bubble up until it reached the `supervisorScope` but will stop there.

**Pull request closed without merging**